sympy/logic/boolalg.py

```diff
@@ -1363,7 +1364,7 @@ def __new__(cls, *args, **options):
             return And(*argset)
         if False in argset:
             argset.discard(False)
-            return And(*[~arg for arg in argset])
+            return And(*(~arg for arg in argset))
```

Comment:
Does this kind of thing actually lead to any speed up?
```python
In [1]: %timeit [].extend([a for a in range(1000)])
44.1 µs ± 107 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [2]: %timeit [].extend(a for a in range(1000))
69 µs ± 1.21 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
I think that the chief difference between a list comprehension and a generator expression in performance terms is that in a generator expression the active frame is switched twice for each item yielded effectively adding the cost of a function call for each item.

Internally a list comprehension
```python
a.extend([e for e in stuff])
```
translates to:
```python
def _tmp():
    res = []
    for e in stuff:
        res.append(e)
    return res

_res = _tmp()
a.extend(_res)
```
whereas a generator expression
```
a.extend(e for e in stuff)
```
translates to
```python
def _tmp():
    for e in stuff:
        yield e

_gen = _tmp()
a.extend(_gen)
```
Now as the `a.extend` method iterates over `_gen` the `_tmp` function is repeatedly suspended and restarted to invoke the code between each pair of yields.

What this means is that actually iterating over a generator expression is slower than iterating over a list comprehension. The advantage of the generator expression is that it is lazy so if there is any possibility to shortcut the iteration then we don't have to compute all of the items and then it can be a big performance boost. Also if the full list of items would e.g. use too much memory or something then the generator makes it possible to process one at a time with `O(1)` memory. 

If all items from the iterable are consumed and ultimately stored in memory simultaneously then the advantages of a generator are gone and it's faster just to use a list comprehension. Here the use of `*` unpacking forces the generator into a tuple to be passed to `And.__new__` I think it's faster to create that tuple from a list than from a generator expression. In any case if we don't get to shortcut then any potential optimisations here are unlikely to be noticeable. I expect that `And.__new__` plus sympifying every element will be much slower.

If we did want to optimise this kind of thing then potentially a good approach is adding an optimised `Add._fromiter` function could work. The idea would be that all inputs are already sympified and the argument is potentially a generator that can be shortcutted. Then `_fromiter` could iterate over the generator looking for `S.false` to exit early. (Note that there is a `Basic.fromiter` method but it has no performance benefits in any situation.)

In any case with optimisations like this we need some reasoning or benchmark or timings to justify them or otherwise it's best just to leave things alone.

Reply:
For the `_convert_to_varsSOP/POS` functions I tested a bit and there it is better to not create a list. Not sure about the other cases, so I'll revert those.